core/templates/pages/exploration-editor-page/services/change-list.service.ts

```diff
@@ -17,246 +17,280 @@
+  }
+
+  /**
+   * Saves a change dict that represents adding a new state. It is the
+   * responsbility of the caller to check that the new state name is valid.
+   *
+   * @param {string} stateName - The name of the newly-added state
+   */
+
+  addState(stateName: string): void {
+    this.addChange({
+      cmd: this.CMD_ADD_STATE,
+      state_name: stateName
+    });
+  }
+
+  /**
+   * Saves a change dict that represents deleting a new state. It is the
+   * responsbility of the caller to check that the deleted state name
+   * corresponds to an existing state.
+   *
+   * @param {string} stateName - The name of the deleted state.
+   */
+
+  deleteState(stateName: string): void {
+    this.addChange({
+      cmd: this.CMD_DELETE_STATE,
+      state_name: stateName
+    });
+  }
+
+  discardAllChanges(): Promise<void> {
+    this.explorationChangeList = [];
+    this.undoneChangeStack = [];
+    return this.explorationDataService.discardDraft();
+  }
+
+  /**
+   * Saves a change dict that represents a change to an exploration
+   * property (such as its title, category, ...). It is the responsibility
+   * of the caller to check that the old and new values are not equal.
+   *
+   * @param {string} backendName - The backend name of the property
+   *   (e.g. title, category)
+   * @param {string} newValue - The new value of the property
+   * @param {string} oldValue - The previous value of the property
+   */
+
+  editExplorationProperty(
+      backendName: string, newValue: string, oldValue: string): void {
+    if (!this.ALLOWED_EXPLORATION_BACKEND_NAMES.hasOwnProperty(backendName)) {
+      this.alertsService.addWarning(
+        'Invalid exploration property: ' + backendName);
+      return;
+    }
+    this.addChange({
+      cmd: this.CMD_EDIT_EXPLORATION_PROPERTY,
+      new_value: angular.copy(newValue),
+      old_value: angular.copy(oldValue),
+      property_name: backendName
+    });
+  }
+
+  /**
+   * Saves a change dict that represents a change to a state property. It
+   * is the responsibility of the caller to check that the old and new
+   * values are not equal.
+   *
+   * @param {string} stateName - The name of the state that is being edited
+   * @param {string} backendName - The backend name of the edited property
+   * @param {string} newValue - The new value of the property
+   * @param {string} oldValue - The previous value of the property
+   */
+
+  editStateProperty(
+      stateName: string, backendName: string,
+      newValue: string, oldValue: string): void {
+    if (!this.ALLOWED_STATE_BACKEND_NAMES.hasOwnProperty(backendName)) {
+      this.alertsService.addWarning('Invalid state property: ' + backendName);
+      return;
+    }
+    this.addChange({
+      cmd: this.CMD_EDIT_STATE_PROPERTY,
+      new_value: angular.copy(newValue),
+      old_value: angular.copy(oldValue),
+      property_name: backendName,
+      state_name: stateName
+    });
+  }
+
+  getChangeList(): Object[] {
```

Comment:
Not Object[] It should be ExplorationChange[] I think

Reply:
done!